src/pointer/ptr.rs

```diff
@@ -158,104 +133,156 @@ mod def {
 
 pub use def::Ptr;
 
-/// The parameterized invariants of a [`Ptr`].
-//
-// TODO(https://github.com/rust-lang/rust/issues/95174): These items are named
-// according to the conventions for enums, should be treated as if they are
-// enums (i.e., only referenced by name), and should be re-defined as enums as
-// soon as it is possible to do so.
-#[allow(non_upper_case_globals)]
-#[doc(hidden)]
-pub mod invariant {
-    /// The aliasing invariant of a [`Ptr`][super::Ptr].
-    ///
-    /// Either:
-    /// - [`AnyAliasing`]
-    /// - [`Shared`]
-    /// - [`Exclusive`]
-    pub type Aliasing = u8;
+/// Used to define the system of [invariants][invariant] of `Ptr`.
+macro_rules! define_system {
+    ($(#[$system_attr:meta])* $system:ident {
+        $($(#[$set_attr:meta])* $set:ident {
+            $( $(#[$elem_attr:meta])* $elem:ident,)*
+        })*
+    }) => {
+        mod sealed {
+            pub trait Sealed {}
 
-    /// No aliasing invariant.
-    pub const AnyAliasing: Aliasing = 1 << 0;
+            impl<$($set,)*> Sealed for ($($set,)*)
+            where
+                $($set: super::$set,)*
+            {}
```

Comment:
We could also just put a single `Sealed` trait in `util::sealed` or something. We already have a top-level `sealed` module which we could move to `util::sealed`.